rust/src/util/types.rs

```diff
@@ -1,55 +1,325 @@
+                .to_owned()
+                .into_boxed_slice();
+
+            NonNull::new(Box::into_raw(bytes).cast())
+        });
+
+        fil_Array { ptr, len: self.len }
+    }
 }
 
-impl Default for fil_GpuDeviceResponse {
+impl<T> Default for fil_Array<T> {
     fn default() -> Self {
-        Self {
-            error_msg: ptr::null(),
-            status_code: FCPResponseStatus::FCPNoError,
-            devices_len: 0,
-            devices_ptr: ptr::null(),
+        Self { ptr: None, len: 0 }
+    }
+}
+
+impl<T: Sized> fil_Array<T> {
+    pub fn is_null(&self) -> bool {
+        self.ptr.is_none() || self.len == 0
+    }
+
+    /// Converts this array into a boxed slice, transferring ownership of the memory.
+    pub fn into_boxed_slice(self) -> Box<[T]> {
+        match self.ptr {
+            None => Box::new([]),
+            Some(ptr) => {
+                let res = unsafe {
+                    Box::from_raw(std::slice::from_raw_parts_mut(
+                        ptr.as_ptr().cast(),
+                        self.len,
+                    ))
+                };
+                // no drop for us
+                std::mem::forget(self);
+                res
+            }
         }
     }
 }
 
-code_and_message_impl!(fil_GpuDeviceResponse);
+// This is needed because of https://github.com/rust-lang/rust/issues/59878
+impl<T: Sized> IntoIterator for fil_Array<T> {
+    type Item = T;
+    type IntoIter = std::vec::IntoIter<T>;
 
-#[no_mangle]
-pub unsafe extern "C" fn fil_destroy_gpu_device_response(ptr: *mut fil_GpuDeviceResponse) {
-    let _ = Box::from_raw(ptr);
+    #[inline]
+    fn into_iter(self) -> Self::IntoIter {
+        Vec::from(self.into_boxed_slice()).into_iter()
+    }
+}
+
+/// Owned, fixed size and heap allocated array of bytes.
+#[allow(non_camel_case_types)]
+pub type fil_Bytes = fil_Array<u8>;
+
+impl<T> From<Vec<T>> for fil_Array<T> {
+    fn from(buf: Vec<T>) -> Self {
+        if buf.is_empty() {
+            return Default::default();
+        }
+        buf.into_boxed_slice().into()
+    }
+}
+
+impl<T: Clone> From<&[T]> for fil_Array<T> {
+    fn from(buf: &[T]) -> Self {
+        if buf.is_empty() {
+            return Default::default();
+        }
+        buf.to_vec().into_boxed_slice().into()
+    }
+}
+
+impl<T> From<Box<[T]>> for fil_Array<T> {
+    fn from(buf: Box<[T]>) -> Self {
+        if buf.is_empty() {
+            return Default::default();
+        }
+        let len = buf.len();
+        let ptr = NonNull::new(Box::into_raw(buf).cast());
+
+        Self { ptr, len }
+    }
+}
+
+impl<T> Drop for fil_Array<T> {
+    fn drop(&mut self) {
+        if let Some(ptr) = self.ptr {
+            let _ = unsafe { Box::from_raw(ptr.as_ptr()) };
+        }
+    }
+}
+
+impl<T> Deref for fil_Array<T> {
+    type Target = [T];
+    fn deref(&self) -> &[T] {
+        match self.ptr {
+            None => unsafe { std::slice::from_raw_parts(ptr::NonNull::dangling().as_ptr(), 0) },
```

Comment:
nit: I _think_ you can write this as `&[]`, right?

Reply:
you wrote this :P, haven't tried it yet

Reply:
fixed